airflow/providers/google/cloud/operators/cloud_build.py

```diff
@@ -15,186 +15,678 @@
+    template_fields = ("project_id", "id_", "gcp_conn_id")
 
-    def _reformat_source(self):
-        self._reformat_repo_source()
-        self._reformat_storage_source()
+    @apply_defaults
+    def __init__(
+        self,
+        id_: str,
+        project_id: Optional[str] = None,
+        retry: Optional[Retry] = None,
+        timeout: Optional[float] = None,
+        metadata: Optional[Sequence[Tuple[str, str]]] = None,
+        gcp_conn_id: str = "google_cloud_default",
+        *args,
+        **kwargs
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.id_ = id_
+        self.project_id = project_id
+        self.retry = retry
+        self.timeout = timeout
+        self.metadata = metadata
+        self.gcp_conn_id = gcp_conn_id
 
-    def _reformat_repo_source(self):
-        if "repoSource" not in self.body["source"]:
-            return
+    def execute(self, context):
+        hook = CloudBuildHook(gcp_conn_id=self.gcp_conn_id)
+        result = hook.cancel_build(
+            id_=self.id_,
+            project_id=self.project_id,
+            retry=self.retry,
+            timeout=self.timeout,
+            metadata=self.metadata,
+        )
+        return MessageToDict(result)
+
+
+class CloudBuildCreateBuildOperator(BaseOperator):
+    """
+    Starts a build with the specified configuration.
 
-        source = self.body["source"]["repoSource"]
+    :param build: The build resource to create. If a dict is provided, it must be of the same form
+        as the protobuf message `google.cloud.devtools.cloudbuild_v1.types.Build`
+    :type build: Union[dict, `google.cloud.devtools.cloudbuild_v1.types.Build`]
+    :param project_id: Optional, Google Cloud Project project_id where the function belongs.
+        If set to None or missing, the default project_id from the GCP connection is used.
+    :type project_id: Optional[str]
+    :param wait: Optional, wait for operation to finish.
+    :type wait: Optional[bool]
+    :param retry: Optional, a retry object used  to retry requests. If `None` is specified, requests
+        will not be retried.
+    :type retry: Optional[Retry]
+    :param timeout: Optional, the amount of time, in seconds, to wait for the request to complete.
+        Note that if `retry` is specified, the timeout applies to each individual attempt.
+    :type timeout: Optional[float]
+    :param metadata: Optional, additional metadata that is provided to the method.
+    :type metadata: Optional[Sequence[Tuple[str, str]]]
+    :param gcp_conn_id: Optional, the connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: Optional[str]
+
+    :rtype: dict
+    """
 
-        if not isinstance(source, str):
-            return
+    template_fields = ("project_id", "build", "gcp_conn_id")
 
-        self.body["source"]["repoSource"] = self._convert_repo_url_to_dict(source)
+    @apply_defaults
+    def __init__(
+        self,
+        build: Union[Dict, Build],
```

Comment:
Can you add backward compatibility with the previous operator?

Reply:
I'm thinking we should keep the parameter name as `body` or use the new name `build`, or use both. @mik-laj

Reply:
WDYT? @mik-laj

Reply:
I was mainly worried about the lack of this logic.
https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/operators/cloud_build.py#L32-L154
It will be best to support two parameters and display a message when the old one is used.

Reply:
@mik-laj Done. PTAL